mempool/clist_mempool.go

```diff
@@ -780,3 +784,42 @@ func (rc *recheck) setRecheckFull() bool {
 func (rc *recheck) consideredFull() bool {
 	return rc.recheckFull.Load()
 }
+
+// CListIterator implements an Iterator that traverses the CList sequentially. When the current
+// entry is removed from the mempool, the iterator starts from the beginning of the CList. When it
+// reaches the end, it waits until a new entry is appended.
+type CListIterator struct {
+	txs    *clist.CList    // to wait on and retrieve the first entry
+	cursor *clist.CElement // pointer to the current entry in the list
+}
+
+func (mem *CListMempool) NewIterator() Iterator {
+	return &CListIterator{
+		txs: mem.txs,
+	}
+}
+
+// WaitNextCh returns a channel to wait for the next available entry. The channel will be closed
+// once the entry is added to the channel or when reaching the end of the list.
+func (iter *CListIterator) WaitNextCh() <-chan Entry {
+	ch := make(chan Entry)
+	// Spawn goroutine that waits for the next entry, saves it locally, and puts it in the channel.
+	go func() {
+		if iter.cursor == nil {
+			// We are at the beginning of the iteration or the saved entry got removed: wait until
+			// the list becomes not empty and select the first entry.
+			<-iter.txs.WaitChan()
+			iter.cursor = iter.txs.Front()
+		} else {
+			// Wait for the next entry after the current one.
+			<-iter.cursor.NextWaitChan()
+			iter.cursor = iter.cursor.Next()
+		}
+		// Check if we reached the end of the list (CElement.Next returned nil).
+		if iter.cursor != nil {
+			ch <- iter.cursor.Value.(Entry)
+		}
+		close(ch)
```

Comment:
Remember that closing a channel also sends a `nil` value 

```
case entry = <-iter.WaitNextCh():
```

will fire with `nil` and we will execute `continue`

Reply:
The receiver will get `nil` when closing the channel without sending any message on it. This happens when `iter.cursor == nil`. When `iter.cursor != nil`, it first send the message on the channel and then it closes it, but the receiver (line 227) already got the message. (I checked this behaviour on the code.)

Reply:
To make it more clear I can change it to:
```
if iter.cursor != nil {
    ch <- iter.cursor.Value.(Entry)
} else {
    close(ch)
}
```